Allow to pass a connection to the `dbconsole` command

robin850 opened the pull request:
Hello,

Since 0a4f6009, it's possible to specify a 3-level database configuration to gather connections by environment.

The `dbconsole` command will try to look for a database configuration which points to the current environment but with such flavour, the environment key is flushed out so let's add the ability to specify the connection and pick `primary` by default to be consistent with Active Record.

Have a nice day !

robin850 committed:
Allow to pass a connection to the `dbconsole` command

Since 0a4f6009, it's possible to specify a 3-level database
configuration to gather connections by environment.

The `dbconsole` command will try to look for a database configuration
which points to the current environment but with such flavour, the
environment key is flushed out so let's add the ability to specify
the connection and pick `primary` by default to be consistent with
Active Record.

y-yagi commented on railties/lib/rails/commands/dbconsole/dbconsole_command.rb:
```diff
@@ -99,6 +99,10 @@ def environment
     end
 
+    def connection
+      @options[:connection] || "primary"
```
Can we use thor's default value?

```ruby 
      class_option :connection, aliases: "-c", type: :string,
        desc: "Specifies the connection to use.", default: "primary"
```

robin850 replied:
Yes, totally fair point and thus, this will be printed running `--help`, thanks !

robin850 replied:
Actually I added it back because `Rails::DBConsole` is public API so people may be using it directly, without Thor. 😊

jeremy commented on railties/lib/rails/commands/dbconsole/dbconsole_command.rb:
```diff
@@ -87,10 +87,10 @@ def start
       @config ||= begin
-        if configurations[environment].blank?
+        if configurations[environment].blank? && configurations[connection].blank?
           raise ActiveRecord::AdapterNotSpecified, "'#{environment}' database is not configured. Available configuration: #{configurations.inspect}"
```
This error will be misleading if the connection was provided.

jeremy commented on railties/lib/rails/commands/dbconsole/dbconsole_command.rb:
```diff
@@ -87,10 +87,10 @@ def start
           raise ActiveRecord::AdapterNotSpecified, "'#{environment}' database is not configured. Available configuration: #{configurations.inspect}"
         else
-          configurations[environment]
+          configurations[environment] || configurations[connection]
```
If `configurations[environment]` is `{}` and `configurations[connection]` is present, we end up returning the blank env config here instead of the conn config.

jeremy commented on railties/lib/rails/commands/dbconsole/dbconsole_command.rb:
```diff
@@ -99,6 +99,10 @@ def environment
+    def connection
+      @options["connection"] || "primary"
+    end
+
```
Nice case for `Hash#fetch`

jeremy commented on railties/lib/rails/commands/dbconsole/dbconsole_command.rb:
```diff
@@ -145,6 +149,9 @@ class DbconsoleCommand < Base # :nodoc:
+        desc: "Specifies the connection to use.", default: "primary"
+
       def perform
         extract_environment_option_from_argument
```
I'd expect to be able to provide the conn as a plain arg. Example scenarios/possibilities:
```bash
rails dbconsole production replica

rails dbconsole production:replica

RAILS_ENV=production rails dbconsole replica

rails dbconsole reporting-replica
```

kaspth replied:
Think we should deprecate the positional environment arg then and have:

```bash
rails dbconsole replica -e production
```

jeremy commented on railties/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+    command when using a 3-level database configuration.
+
+    *Robin Dupret*
+
```
Let's show an example of this invocation style, too 👍

jeremy reviewed (commented):
Yes please!

robin850 committed:
Deprecate environment as an argument for dbconsole and console

People should rather rely on the `-e` or `--environment` options to
specify in which environment they want to work. This will allow us
to specify the connection to pick as a regular argument in the future.

robin850 commented:
Jeremy, there is now a `presence` check in case `configurations[environment]` is `{}` but I don't see how that could happen because this would've been catch by `blank?` above. Also, with the new configuration system, Active Record is either returning a) a configuration with the environment's as a key (2-level db config) or b) a configuration without any occurrence of the current environment but rather a key for each possible connection (3-level db config). We can keep it but I'm not sure this is useful.

Kasper, passing the environment as a regular argument is deprecated as you suggested ; this is certainly going to be easier for us to handle only the connection as a regular argument. Such flavor is not deprecated for the `console` command as I don't know whether we are going to open all available connections by default (c.f. https://github.com/rails/rails/pull/29360) or let people open them manually (thus, we would have to do that for `console` as well).

Finally, there's a tiny fix attached with this pull request because if we are going to deprecate passing the environment as a regular argument, people may encounter issues as the `-e` notation didn't expand shortcuts (like `dev` for `devlopment`).

Thanks for the feedback guys ! :-)

kaspth commented:
Can you rebase? The changelog conflicts.

I think we should deprecate the environment positional command argument for `console` as well for consistency.

robin850 committed:
Properly expand the environment's name

Running the `console` and `dbconsole` commands with a regular argument
as the environment's name automatically expand it to match an existing
environment (e.g. dev for development).

This feature wasn't available using the `--environment` (a.k.a `-e`)
option.

robin850 commented:
@kaspth : Rebased and deprecated the environment argument for the `console` command.

kaspth commented on railties/lib/rails/command/environment_argument.rb:
```diff
@@ -7,13 +7,24 @@ module EnvironmentArgument #:nodoc:
             self.options = options.merge(environment: acceptable_environment(environment))
-          elsif !options[:environment]
+
+            ActiveSupport::Deprecation.warn "Passing the environment's name as a " \
```
Much cleaner to have this straight in here 👏
